Fix and improve ImageSet intersection handler

gschintgen opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #18115 
Fixes #18196
Improves #17638 (no longer raises; `solve` returns correct result, but actual `ImageSet` intersection not computed)
Improves #18081 (no longer raises)
Improves #17568 (no longer raises)
Improves #9616 (no longer raises)

#### Brief description of what is fixed or changed
Fix diophantine() issue #18196 
In some cases negative coefficients could lead to `classify_diop` not recognizing the equation.
Thanks to @smichr for the fix and the test.

This PR also fixes and improves `ImageSet`'s `diophantine`-related intersection handler:
- `ImageSet` intersection no longer raises an exception when diophantine can't solve an equation.
- Correctly handle the case where both `ImageSets` use the same variable in their respective `Lambdas`.
- Correctly handles the various possible solution sets returned by `diophantine`.
- `diophantine()` must be asked explicitely to return all solutions using `permute=True`;

Tests are added.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
  * Fixed a bug in `diophantine` where negative coefficients could lead to incomplete solutions in conjunction with `permute=True`.
* sets
  * Fixed and extended intersection handler for `Integer`-based `ImageSets`.
<!-- END RELEASE NOTES -->

gschintgen committed:
Fix diophantine() issue #18196

In some cases negative coefficients could lead to classify_diop
not recognizing the equation.

Thanks to Christopher Smith @smichr for the fix and the test.

gschintgen committed:
Fix and improve ImageSet intersection's diophantine code

Fixes #18115: ImageSet intersection no longer raises an
exception when diophantine can't solve an equation.

This commit also fixes various other errors in the
diophantine-calling code:
- correctly handle the case where both ImageSets use
  the same variable in their respective Lambdas;
- correctly handles the various possible solution sets
  returned by diophantine;
- diophantine() must be asked explicitely to return all
  solutions using permute=True;
- adds tests for the relevant cases.

oscarbenjamin commented on sympy/sets/handlers/intersection.py:
```diff
@@ -235,26 +235,44 @@ def intersection_sets(self, other): # noqa:F811
+            try:
+                solns = list(diophantine(fn - gm, syms=(n, m), permute=True))
+            except (TypeError, NotImplementedError):
+                return
```
What raises `TypeError`?

gschintgen replied:
https://github.com/sympy/sympy/blob/c559a8421ac4865ebfe66024be6cd43a6103a62b/sympy/solvers/diophantine.py#L202-L204
That's the exception raised by `diophantine` that was at the core of a number of filed issues. `diophantine` does a good job of checking if it can handle the passed equation and this is how it complains about one specific type of "misuse". (Whereas `NotImplementedError` is raised if the equation would be fine, but it simply doesn't know how to solve it.)

oscarbenjamin replied:
I think that should be `ValueError` rather than `TypeError`. I also very much dislike the way that the diophantine function uses large try blocks. Catching `AttributeError` shouldn't happen anywhere...

oscarbenjamin replied:
I just tried this PR with this diff:
```diff
diff --git a/sympy/sets/handlers/intersection.py b/sympy/sets/handlers/intersection.py
index 848cf2ee10..55c7ec35bd 100644
--- a/sympy/sets/handlers/intersection.py
+++ b/sympy/sets/handlers/intersection.py
@@ -252,7 +252,7 @@ def intersection_sets(self, other): # noqa:F811
             n = self.lamda.variables[0]
             try:
                 solns = list(diophantine(fn - gm, syms=(n, m), permute=True))
-            except (TypeError, NotImplementedError):
+            except NotImplementedError:
                 return
             # 3 cases are possible for solns:
             # - empty set,
```
I ran all non-slow tests in sets and solvers and didn't get any failures...

gschintgen replied:
> I also very much dislike the way that the `diophantine` function uses large try blocks

Agreed. I was astonished to see such a large `try` block when I investigated #18196 a bit. (And that `try` block even contributed to it.)

As for the `TypeError`, it should be sufficient to try the one from #18018 
```
imageset(Lambda(n, n*log(2)), S.Integers).intersection(S.Integers)
```

gschintgen replied:
I'll add a test, simply to exclude that this execption is removed from the list.

oscarbenjamin replied:
Can you also add a comment explaining where the exceptions are raised?

The biggest problem with cleaning up all the exception around the codebase is not knowing what it potentially catches or what the author intended it to catch.

gschintgen replied:
Sure, I'll add a note.

oscarbenjamin replied:
As far as I can tell from the tests the `AttributeError` is just there to handle this case:
```julia
In [1]: diophantine(3)                                                                                                            
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/current/sympy/sympy/sympy/solvers/diophantine.py in diophantine(eq, param, syms, permute)
    174     try:
--> 175         var = list(eq.expand(force=True).free_symbols)
    176         var.sort(key=default_sort_key)

AttributeError: 'int' object has no attribute 'expand'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-1-50971d7489ec> in <module>
----> 1 diophantine(3)

~/current/sympy/sympy/sympy/solvers/diophantine.py in diophantine(eq, param, syms, permute)
    201         assert eq.is_polynomial()
    202     except (GeneratorsNeeded, AssertionError, AttributeError):
--> 203         raise TypeError(filldedent('''
    204     Equation should be a polynomial with Rational coefficients.'''))
    205 

TypeError: 
Equation should be a polynomial with Rational coefficients.
```
However the sympified version works fine:
```julia
In [3]: diophantine(S(3))                                                                                                         
Out[3]: set()
```
The error is just that diophantine doesn't `eq = _sympify(eq)` so the only effect of catching `AttributeError` is to cover up that bug.

gschintgen commented on sympy/sets/tests/test_fancysets.py:
```diff
@@ -618,6 +618,37 @@ def test_imageset_intersect_interval():
+def test_imageset_intersect_diophantine():
+    #from sympy.abc import n, m
+    m = Symbol('m', integer=True)
+    n = Symbol('n', integer=True)
```
Oops, forgot to clean up.

gschintgen committed:
Add tests and notes for ImageSet intersection handler

- exception handling documented
- tests added
- (very minor cleanup of spurious empty lines)

oscarbenjamin commented:
Looks good!
